### PR TITLE
Improve transformer api

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
Small PR allowing the chaining behavior of `TransformerBasis`.

Before this PRs, basis methods returning `self` could not be chained when the basis was wrapped by the `TransformerBasis` class.

Old behavior

```python
>>> import nemos as nmo
>>> from nemos.basis._basis import Basis
>>> from nemos.basis._transformer_basis import TransformerBasis
>>> transformer_basis = nmo.basis.BSplineEval(5).to_transformer()
>>> out = transformer_basis.set_input_shape(1)
>>> isinstance(out, TransformerBasis)
False
>>> isinstance(out, Basis)
True
```

New behavior

```python
>>> import nemos as nmo
>>> from nemos.basis._basis import Basis
>>> from nemos.basis._transformer_basis import TransformerBasis
>>> transformer_basis = nmo.basis.BSplineEval(5).to_transformer()
>>> out = transformer_basis.set_input_shape(1)
>>> isinstance(out, TransformerBasis)
True
>>> isinstance(out, Basis)
False
```

## Key Features
- Decorator that processes the output of chainable basis methods, setting the method output (the updated basis object) to the `_basis` attribute of the transformer, and returns the transformer.
- New tuple attribute of the `TransformerBasis` listing all the chainable methods of basis.
- Run-time decoration of chainable methods in the `__getattr__` when first called + caching. Run-time decorating is necessary because if set at initialization, it would create a circular reference, and result in infinite loop when pickling.

### CODE REVIEW

**Review by sjvenditto — APPROVED** [review]
I committed a fix to a couple typos I found, otherwise it looks good!

**Comment by BalzaniEdoardo** [comment]
Great! Thanks

**Review by billbrod — CHANGES_REQUESTED** [review]
Don't really have any notes about the chaining, that all looks good to me.

- `TransformerBasis` objects have a `to_transformer()` method, which I think just returns self. This is a bit confusing -- is there a reason to have this behavior?
- `transform` and `fit_transform` should have the notes about `set_input_shape` being called first as well (like `fit`), as well as the Raises section of the docstring

File: src/nemos/basis/_transformer_basis.py [hunk 1]
```diff
@@ -112,7 +133,11 @@ def fit(self, X: FeatureMatrix, y=None):
         """
         Compute the convolutional kernels.
 
-        If any of the 1D basis in self._basis is in "conv" mode, it computes the convolutional kernels.
+        Checks the input structure and, if any of the 1D basis in self._basis is in "conv" mode,
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/basis/_transformer_basis.py:136]: Yes, I'll remove the properties

File: src/nemos/basis/_transformer_basis.py [hunk 2]
```diff
@@ -286,21 +340,21 @@ def __setattr__(self, name: str, value) -> None:
...
+        ...     trans_bas.rand_atrr = "some value"
         ... except ValueError as e:
         ...     print(repr(e))
-        ValueError('Only setting _basis or existing attributes of _basis is allowed.')
+        ValueError('Only setting _basis or existing attributes of _basis is allowed. Attempt to set `rand_atrr`.')
         """
-        # allow self._basis = basis
-        if name == "_basis":
+        # allow self._basis = basis and other attrs of self to be retrievable
+        if name in ["_basis", "basis", "_wrapped_methods"]:
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/basis/_transformer_basis.py:351]: I did remove the property and store directly basis in `self.basis`. We need it writable to cross validate the basis directly

File: tests/test_transformer_basis.py [hunk 3]
```diff
@@ -0,0 +1,898 @@
...
+
+import nemos as nmo
+from nemos import basis
+from nemos._inspect_utils import get_subclass_methods, list_abstract_methods
+from nemos.basis import AdditiveBasis, MSplineConv, MultiplicativeBasis
+
+
+@pytest.mark.parametrize(
+    "basis_cls",
+    list_all_basis_classes(),
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply tests/test_transformer_basis.py:18]: yes, exactly, the composite bases are included otherwise

**Pull request merged**